### ISSUE TITLE
Better development tools

### ISSUE THREAD

**Issue by silentrob** [silentrob/superscript#200]
The editor had some nice topic and gambit matching tools, it would be nice to have something similar from the command line.

Perhaps verifying an input matches (and which gambit/topic)
Verify the conversation threads, or reply chain for more complicated replies.
Test plugins outside of the bot engine.

**Comment by dcsan** [comment]
is the approach of using an editor out of the question entirely for some reason? or is it just temporarily on ice as its too much to take on as well as developing the main system right now?

**Comment by silentrob** [comment]
In terms of roadmap those are one in the same. I don't have any plans in the next 6 months to continue development on the editor, and I don't want to speculate after that.

**Comment by mariusursache** [comment]
I think that an editor, while a nice interface, slows down a lot writing gambits/topics, compared to a file-based approach. If it weren’t (still) a one-man effort, it might be worth doing a more usable editor —maybe even one with a visual decision tree, such as [this one](http://4.bp.blogspot.com/-5P1x7HMJOo8/T98iBrRE0DI/AAAAAAAAAFs/-0ll1WHKD68/s640/GamePartconvo.jpg).

But a tool to check/test more complex conversations would be useful—to avoid running complex scenarios manually (which often might skip cases, besides being time consuming).

**Comment by dcsan** [comment]
I also find a text editor to be the best, but for "authors/writers" as opposed to developers, it's sometimes a bit daunting esp at first.

I really like Twine UI for branching stories, but it's basically just for links and not user freetext input - like a wiki which generates a sitemap in realtime.
![image](https://cloud.githubusercontent.com/assets/514002/11693687/1d0c5406-9e5b-11e5-890e-c7724c2420d2.png)

Separately the cleverscript spreadsheet approach also seems like it would help people get started as there's a clear "place for everything", boxes you fill it and you can also see an overview of the dialog. Very easy to get going with.

but i found spreadsheets get very hard to follow once you have a decent amount of content. perhaps filter views could help that, like excel/google docs has...

http://www.cleverscript.com/about/
![cleverscript_example](https://cloud.githubusercontent.com/assets/514002/11693741/6c7e66a0-9e5b-11e5-81b5-e4dac21979d7.jpg)

very interested in this area, if others have examples or experience. 

I think in terms of IVR authoring systems there must be tons of work in this area. 

using joynt.js
http://etodd.io/2014/05/16/the-poor-mans-dialogue-tree/

unity tools
http://forum.unity3d.com/threads/released-dialogue-system-for-unity-easy-conversations-quests-and-more.204752/


Also had some discussion on this for Rivescript editor
https://github.com/aichaos/rivescript-js/issues/22

**Comment by Lewwwk** [comment]
In ideal world nice to have, and good for broader adoption, but feels low priority to me and possibly better handled through a separate project like @neoatbay was suggesting in the slack channel.

**Comment by silentrob** [comment]
Yes I agree.. which was why the editor was a separate project. Im just thinking about added a few bare bones tools to improve or help debugging and improve the authoring experience a little.

**Comment by Lewwwk** [comment]
Would be useful - and increasingly so as scripts and use cases grow.

**Comment by ganarajpr** [comment]
I think debugging tools will be extremely useful - whether its for authors or developers.. When developing a long conversational tree with multiple topics and branches something like that would be extremely useful.. 

I think @silentrob's editor has the right direction. It probably needs a good look at usability from a newb point of view and it would be great.

**Comment by dcsan** [comment]
in terms of debugging, just a visualisation of the internal state, and how it got there would be very useful.
ie a read-only view into the bots decisions. eg the current variables and a history of why it changed to topicX. since superscript is designed to be less deterministic than say Rivescript, it maybe sometimes perplexing why the conversation flowed where it did...

where editors get much more complex is if its an authoring tool, ie read and write not just a viewer.
